Preserve leading whitespace for Normal lines

gfontenot opened the pull request:
Sometimes people use leading whitespace for formatting reasons when 
composing plain text emails. It's a shame to have people go through that 
trouble just to have us strip that formatting when we go reflow the text.

To fix this, we can add an indentation level to the Normal constructor, and
take any whitespace we find into account. Then we can indent the whole line
(and all wrapped lines resulting from that line) with the existing
indentation level to maintain formatting.

This actually breaks our existing functionality of treating any
whitespace only lines as a newline, so I also went ahead and refactored
that to be more explicit.

gfontenot committed:
Preserve leading whitespace for Normal lines

Sometimes people use leading whitespace for formatting reasons when
composing plain text emails. It's a shame to have people go through that
trouble just to have us strip that formatting when we go reflow the
text.

To fix this, we can add an indentation level to the Normal constructor,
and take any whitespace we find into account. Then we can indent the
whole line (and all wrapped lines resulting from that line) with the
existing indentation level to maintain formatting.

Note that this breaks our existing whitespace tests (treating
whitespace-only lines as a newline). That will be fixed again in the
next commit.

gfontenot committed:
Explicitly parse blank lines

Blank lines are important for formatting and we want to parse them
accordingly. By explicitly parsing these, our intentions are a bit
clearer than they were before when we were just treating these lines
like Normal lines.

gfontenot committed:
Refactor prepending code

The old way of prepending was pretty gross and hard to follow. I think
this makes it clearer.

pbrisbin commented on src/Main.hs:
```diff
@@ -41,3 +42,9 @@ splitWordsAt w xs x
 appendToLast xs x = T.unwords [last xs, x]
+
+mapPrepend :: Text -> [Text] -> [Text]
+mapPrepend t ts = map (mappend t) ts
```
You can drop the `ts` from both sides.

pbrisbin commented on src/Reflow/Parser.hs:
```diff
@@ -22,12 +22,20 @@ parseFile t = either (return . parserError t) id
+
+blank :: Parser Content
+blank = do
+    _ <- try $ trailingWhitespace >> eol
```
`Control.Monad.void`?

pbrisbin commented on src/Reflow/Parser.hs:
```diff
@@ -22,12 +22,20 @@ parseFile t = either (return . parserError t) id
+blank :: Parser Content
+blank = do
+    _ <- try $ trailingWhitespace >> eol
+    return Blank
```
FWIW: there's a way to do this "do a thing, discard that and return  this other thing" pattern, I think it's like `try trailingWhitespace >> eol *> Blank` or something.

gfontenot replied:
Yeah, it needs some parens and a `return` as well:

```
blank = (try $ trailingWhitespace >> eol) *> return Blank
```
I kinda find that harder to read than this, so I think I'm going to leave it this way (using `void`).

pbrisbin commented on src/Reflow/Types.hs:
```diff
@@ -11,5 +11,5 @@ data Config = Config
 lessWidth i c = c { width = width c - i }
 
-data Content = Normal Text | Header Text | Quoted Content | CodeBlock Text | PGPBlock Text
+data Content = Normal Int Text | Blank | Header Text | Quoted Content | CodeBlock Text | PGPBlock Text
```
For some reason, my brain wants `Blank` first, funny.

gfontenot replied:
yeah, me too actually.

gfontenot committed:
Small refactorings

 - Closer to point-free for `mapPrepend`
 - use `void` for ignored parsing
 - move Blank to the beginning of the constructors because that just
   kinda makes more sense for some reason.
